service/cachemdw/middleware.go

```diff
@@ -0,0 +1,109 @@
+package cachemdw
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+
+	"github.com/kava-labs/kava-proxy-service/decode"
+)
+
+type cacheContextKey string
+
+const (
+	CachedContextKey cacheContextKey = "X-KAVA-PROXY-CACHED"
+
+	CacheHeaderKey       = "X-Cache"
+	CacheHitHeaderValue  = "HIT"
+	CacheMissHeaderValue = "MISS"
+)
+
+func (c *ServiceCache) Middleware(
+	next http.Handler,
+) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		uncachedContext := context.WithValue(r.Context(), CachedContextKey, false)
+
+		req := r.Context().Value(c.decodedRequestContextKey)
+		decodedReq, ok := (req).(*decode.EVMRPCRequestEnvelope)
+		if !ok {
+			next.ServeHTTP(w, r.WithContext(uncachedContext))
+			return
+		}
```

Comment:
ah, i see you are using the decoded request.

can you help me understand conceptually how this is working? i am under the impression that, in the proxy's current form, the request is decoded only _after_ the response is given (the decoded data is only used for metric generation).

if that's the case, then we'll never get a cache hit because the decoded request will never be in the context. the decoding needs to be moved into the critical path before the response is made.

Reply:
the request is decoded in the `RequestLoggingMiddleware`:
https://github.com/Kava-Labs/kava-proxy-service/blob/69b86f0b542973522e729d8633d3291fb7bac192/service/middleware.go#L124

ah i was misreading this.. it's decoded and then served with the passed in handler which is the proxy:
https://github.com/Kava-Labs/kava-proxy-service/blob/69b86f0b542973522e729d8633d3291fb7bac192/service/service.go#L46

so request is decoded & put in context, then served by the proxy. so it will indeed exist. well that's helpful! i was taking the wrong path for what i'm working on 😅 thanks for being my rubber ducky.